plugins/janus_nosip.c

```diff
@@ -1760,6 +1770,72 @@ char *janus_nosip_sdp_manipulate(janus_nosip_session *session, janus_sdp *sdp, g
 	return janus_sdp_write(sdp);
 }
 
+static int janus_nosip_bind_socket(int fd, int port) {
+	struct sockaddr_in rtp_address;
+	rtp_address.sin_family = AF_INET;
+	rtp_address.sin_port = htons(port);
+	inet_pton(AF_INET, local_ip, &rtp_address.sin_addr.s_addr);
+	if(bind(fd, (struct sockaddr *)(&rtp_address), sizeof(struct sockaddr)) < 0) {
+		JANUS_LOG(LOG_VERB, "Bind failed (port %d)\n", port);
+		return -1;
+	}
+	return 0;
+}
+
+/* Bind RTP/RTCP port pair */
+static int janus_nosip_allocate_port_pair(int fds[2], int ports[2]) {
+	uint16_t rtp_port_next = rtp_range_slider; 					/* Read global slider */
+	uint16_t rtp_port_start = rtp_port_next;
+	gboolean rtp_port_wrap = FALSE;
+
+	int rtp_fd = -1, rtcp_fd = -1;
+	while(1) {
+		if(rtp_port_wrap && rtp_port_next >= rtp_port_start) {	/* Full range scanned */
+			JANUS_LOG(LOG_ERR, "No ports available for audio channel in range: %u -- %u\n",
+					  rtp_range_min, rtp_range_max);
+			break;
+		}
+		if(rtp_fd == -1) {
+			rtp_fd = socket(AF_INET, SOCK_DGRAM, 0);
+		}
+		if(rtcp_fd == -1) {
+			rtcp_fd = socket(AF_INET, SOCK_DGRAM, 0);
+		}
+		if(rtp_fd == -1 || rtcp_fd == -1) {
+			JANUS_LOG(LOG_ERR, "Error creating audio sockets...\n");
+			break;
+		}
+	 	int rtp_port = rtp_port_next;
+		int rtcp_port = rtp_port+1;
+		if((uint32_t)(rtp_port_next + 2UL) < rtp_range_max) {	/* Advance to next pair */
+			rtp_port_next += 2;
+		} else {
+			rtp_port_next = rtp_range_min;
+			rtp_port_wrap = TRUE;
+		}
+		if(janus_nosip_bind_socket(rtp_fd, rtp_port)) { 
+			/* rtp_fd still unbind, reuse it */
+		} else if(janus_nosip_bind_socket(rtcp_fd, rtcp_port)) {
+			close(rtp_fd);
+			rtp_fd = -1;
+			/* rtcp_fd still unbind, reuse it */
+		} else {
+			fds[0] = rtp_fd;
+			fds[1] = rtcp_fd;
+			ports[0] = rtp_port;
+			ports[1] = rtcp_port;
+			rtp_range_slider = rtp_port_next; 					/* Write global slider */
```

Comment:
I still haven't understood what this slider is for, but you're modifying it in a non-thread safe way.

Reply:
This is just a helper which points to a last position available. There is no need to synchronize access across different threads since the only drawback is that will have to check a couple more port candidates.